src/assets/TokenDetectionController.test.ts

```diff
@@ -348,119 +315,204 @@ describe('TokenDetectionController', () => {
+    expect(tokensController.state.detectedTokens).toStrictEqual([]);
   });
 
-  it('should not detect tokens if there is no selectedAddress set', async () => {
-    tokenDetection.configure({ networkType: MAINNET });
+  it('should detect new tokens after switching between accounts', async () => {
+    sinon
+      .stub(tokensController, '_instantiateNewEthersProvider')
+      .callsFake(() => null);
+
+    preferences.setSelectedAddress('0x0001');
     getBalancesInSingleCall.resolves({
-      '0x514910771af9ca656af840dff83e8264ecf986ca': new BN(1),
+      [sampleTokenA.address]: new BN(1),
     });
+    await tokenDetection.start();
+    expect(tokensController.state.detectedTokens).toStrictEqual([sampleTokenA]);
+
+    preferences.setSelectedAddress('0x0002');
     await tokenDetection.detectTokens();
-    expect(tokensController.state.tokens).toStrictEqual([]);
+    expect(tokensController.state.detectedTokens).toStrictEqual([sampleTokenA]);
   });
 
-  it('should subscribe to new sibling detecting tokens when account changes', async () => {
+  it('should not detect tokens after stopping polling, and then switching between networks that support token detection', async () => {
     sinon
       .stub(tokensController, '_instantiateNewEthersProvider')
       .callsFake(() => null);
-    const firstNetworkType = 'rinkeby';
-    const secondNetworkType = 'mainnet';
-    const firstAddress = '0x123';
-    const secondAddress = '0x321';
-    const detectTokens = sinon.stub(tokenDetection, 'detectTokens');
-    preferences.update({ selectedAddress: secondAddress });
-    preferences.update({ selectedAddress: secondAddress });
-    expect(preferences.state.selectedAddress).toStrictEqual(secondAddress);
-    expect(detectTokens.calledTwice).toBe(false);
-    preferences.update({ selectedAddress: firstAddress });
-    expect(preferences.state.selectedAddress).toStrictEqual(firstAddress);
+
+    tokenDetection.configure({
+      selectedAddress: '0x1',
+    });
+    const detectedTokensMock = sinon
+      .stub(tokenDetection, 'detectTokens')
+      .callsFake(() => Promise.resolve());
+
     network.update({
       provider: {
-        type: secondNetworkType,
-        chainId: NetworksChainId[secondNetworkType],
+        type: 'rpc',
+        chainId: '56',
       },
     });
-    expect(network.state.provider.type).toStrictEqual(secondNetworkType);
+
+    await tokenDetection.start();
+    expect(detectedTokensMock.callCount).toBe(2);
+
+    tokenDetection.stop();
     network.update({
       provider: {
-        type: firstNetworkType,
-        chainId: NetworksChainId[firstNetworkType],
+        type: 'rpc',
+        chainId: '137',
       },
     });
-    expect(network.state.provider.type).toStrictEqual(firstNetworkType);
-    tokensController.update({ tokens: TOKENS });
-    expect(tokenDetection.config.tokens).toStrictEqual(TOKENS);
+
+    // Detected tokens will still be called once more but will not actually detect tokens
```

Comment:
This assertion is no longer testing that this function behaves correctly, now that `detectTokens` is called either way. This pattern of mocking out a method for test purposes isn't recommended anyway, because now the test is more fragile as it depends upon the behavior of the method under test _and_ the mocked out method. Ideally unit tests should make refactoring easier, but this style of test makes it more difficult.

Perhaps this assertion could be replaced by one that ensure `getBalancesInSingleCall` wasn't called?

Reply:
Updated to test for `getBalancesInSingleCall` instead